### PR TITLE
Share SourceWriter between JSON & config binding generators

### LINKED ISSUE #88918: `.\build.cmd -subset clr+libs -config release` fails in `ConfigurationBindingGenerator`
As per the title, currently running `.\build.cmd -subset clr+libs -config release` currently fails with the following:
```
CSC : error CS8785: Generator 'ConfigurationBindingGenerator' failed to generate source. It will not contribute to the
output and compilation errors may occur as a result. Exception was of type 'ArgumentOutOfRangeException' with message '
repeatCount ('-4') must be a non-negative value. (Parameter 'repeatCount') [D:\Users\tagoo\source\repos\runtime\src\lib
raries\Microsoft.Extensions.Logging.Configuration\src\Microsoft.Extensions.Logging.Configuration.csproj::TargetFramewor
k=net8.0]
D:\Users\tagoo\source\repos\runtime\src\libraries\Microsoft.Extensions.Logging.Configuration\src\LoggerFilterConfigureO
ptions.cs(54,72): error IL2026: Using member 'Microsoft.Extensions.Configuration.ConfigurationBinder.GetValue<T>(IConfi
guration, String, T)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application c
ode. In case the type is non-primitive, the trimmer cannot statically analyze the object's type so its members may be t
rimmed. [D:\Users\tagoo\source\repos\runtime\src\libraries\Microsoft.Extensions.Logging.Configuration\src\Microsoft.Ext
ensions.Logging.Configuration.csproj::TargetFramework=net8.0
```

### PR DESCRIPTION
**Pull request by layomia**
Fixes https://github.com/dotnet/runtime/issues/88918. FYI @tannergooding @ericstj @jasper-d @martincostello @vcsjones @lateralusX.

The writer used for the JSON generator has the correct right line-ending behavior.

### CODE REVIEW

File: src/libraries/Common/src/SourceGenerators/SourceWriter.cs [hunk 1]
```diff
@@ -85,8 +64,18 @@ public void WriteLine(string text)
         public SourceText ToSourceText()
         {
             Debug.Assert(_indentation == 0 && _sb.Length > 0);
-            return SourceText.From(_sb.ToString(), Encoding.UTF8);
+            return SourceText.From(ToString(), Encoding.UTF8);
+        }
+
+        public override string ToString() => _sb.ToString();
+
+        #region Test utilities.
```
Comment (eiriktsarpalis) [inline src/libraries/Common/src/SourceGenerators/SourceWriter.cs:72]: I think we can get rid of the region declaration

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Helpers/Emitter/ConfigurationBinder.cs [hunk 2]
```diff
@@ -25,13 +25,13 @@ private void EmitBinder_ConfigurationBinder()
...
                 _emitBlankLineBeforeNextStatement = false;
-                EmitRootBindingClassBlockStart(Identifier.GeneratedConfigurationBinder);
+                EmitRootBindingClassStartScope(Identifier.GeneratedConfigurationBinder);
 
                 EmitGetMethods();
                 EmitGetValueMethods();
                 EmitBindMethods_ConfigurationBinder();
 
-                _writer.WriteBlockEnd();
+                EmitEndScope();
```
Comment (eiriktsarpalis) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Helpers/Emitter/ConfigurationBinder.cs:34]: Nit: blocks define a new scope but not all scopes are defined using blocks. Perhaps use "block" to emphasize that all the method is doing is emit C-style block syntax.

**Review by eiriktsarpalis — APPROVED** [review]

File: src/libraries/Common/tests/TestUtilities/System/LineEndingsHelper.cs [hunk 3]
```diff
@@ -5,14 +5,9 @@ namespace System
...
-        public const string CompiledNewline = @"
+        private const string CompiledNewline = @"
 ";
-        public static readonly bool s_consistentNewlines = StringComparer.Ordinal.Equals(CompiledNewline, Environment.NewLine);
-
-        static public bool IsNewLineConsistent
-        {
-            get { return s_consistentNewlines; }
-        }
+        private static readonly bool s_consistentNewlines = StringComparer.Ordinal.Equals(CompiledNewline, Environment.NewLine);
```
Comment (ericstj) [inline src/libraries/Common/tests/TestUtilities/System/LineEndingsHelper.cs:10]:
```suggestion
        private static readonly bool s_consistentNewlines = StringComparer.Ordinal.Equals(CompiledNewline, Environment.NewLine);
        
        static public bool IsNewLineConsistent
        {
            get { return s_consistentNewlines; }
        }
```
Looks like this removal broke other tests consuming this file.

**Comment by eiriktsarpalis** [comment]
Shouldn't this be merged before initiating a backport PR?

**Comment by ericstj** [comment]
> Shouldn't this be merged before initiating a backport PR?

We're close on time -- it's fine to merge in parallel so long as we squash the commits in release.

**Comment by layomia** [comment]
Build failure appears unrelated.

**Pull request merged**